MaxText/layers/mistral.py

```diff
@@ -0,0 +1,195 @@
+  mesh: Mesh
+
+  @nn.compact
+  def __call__(self,
+               inputs,
+               decoder_segment_ids,
+               decoder_positions,
+               deterministic,
+               model_mode,
+               ):
+    cfg = self.config
+    mesh = self.mesh
+
+    inputs = nn.with_logical_constraint(
+        inputs, ('activation_batch', 'activation_length', 'activation_embed'))
+
+    lnx_rms = models.RMSNorm(
+        dtype=cfg.dtype,
+        name='pre_self_attention_layer_norm',
+        kernel_axes=('embed',),
+        epsilon=cfg.normalization_layer_epsilon
+        )
+    lnx = lnx_rms(inputs)
+
+    lnx = nn.with_logical_constraint(
+        lnx, ('activation_batch', 'activation_length', 'activation_embed'))
+
+    # Self-attention block
+    attention_layer = Attention(
+      config = cfg,
+      num_query_heads=cfg.num_query_heads,
+      num_kv_heads=cfg.num_kv_heads,
+      head_dim=cfg.head_dim,
+      max_target_length=cfg.max_target_length,
+      attention_kernel=cfg.attention,
+      mesh=mesh,
+      dtype=cfg.dtype,
+      dropout_rate=cfg.dropout_rate,
+      name='self_attention',
+      use_int8=cfg.int8_training)
+
+    attention_lnx = attention_layer(
+            lnx,
+            lnx,
+            decoder_positions,
+            decoder_segment_ids=decoder_segment_ids,
+            deterministic=deterministic,
+            model_mode=model_mode)
+
+    attention_lnx = nn.with_logical_constraint(
+        attention_lnx,
+        ('activation_batch', 'activation_length', 'activation_embed'))
+    intermediate_inputs = inputs + attention_lnx
+
+    # Fully Connected
+    hidden_states = models.RMSNorm(
+        dtype=cfg.dtype, name='post_self_attention_layer_norm', kernel_axes=('embed',),
+        epsilon=cfg.normalization_layer_epsilon,
+        )(intermediate_inputs)
+    hidden_states = nn.with_logical_constraint(hidden_states, ('activation_batch', 'activation_length', 'activation_embed'))
+
+    if cfg.num_experts > 1:
+        # TODO(ranran): currently, this MoeBlock does not work as expected, and plan to fix it in coming PR.
+    
+        # mlp_lnx = linears.MoeBlock(
+        #     config=cfg,
+        #     num_experts=cfg.num_experts,
+        #     num_experts_per_tok=cfg.num_experts_per_tok,
+        #     kernel_init=initializers.nd_dense_init(1.0, 'fan_in', 'truncated_normal'),
+        #     kernel_axes=('embed', 'mlp'),
+        #     name='moe',
+        #     dtype=cfg.dtype,
+        # )(hidden_states, deterministic=deterministic)
+
+        gate_logits = linears.DenseGeneral(
+            cfg.num_experts,
+            dtype=cfg.dtype,
+            kernel_init=initializers.nd_dense_init(
+                1.0, 'fan_in', 'truncated_normal'),
+            kernel_axes=('embed', 'mlp'),
+            name="gate",
+            use_int8=cfg.int8_training,
+        )(hidden_states)
+        weights, selected_experts = jax.lax.top_k(
+            gate_logits, cfg.num_experts_per_tok)
```

Comment:
after this line, adds 
weights = weights.astype(jnp.float32)

Reply:
this is corresponding to this line: https://github.com/mistralai/mistral-src/blob/8598cf582091a596671be31990448e0620017851/mistral/moe.py#L27

Reply:
Updated. Thank you!